### PR TITLE
Fixes on BERT pre-training; Implementing whole-word-masking.

### PR DESCRIPTION
**Pull request by khcs**
[WIP] because didn't do CI testing.

1. BERT pre-training data layer bug-fixes:
- fixing sentence separation
- fixing error on larger seq_length, e.g., max_seq_length=512
- implementing random 'short_len_seq' sampling, as in original BERT
- handling shorter length sequences - proper input_mask
- fixing masking - not masking 'special tokens'
- fixing redundent '[CLS]' and '[SEP]' special token insertions

2. Dockerfile
- adding some NLP-related Python packages in pip-install

3. Implementing whole-word-masking.

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 1]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
+                document = self.tokenizer.text_to_ids(doc_text)
 
-            assert len(document) >= self.max_seq_length
             return document
 
-        a_document = get_document(a_filename, a_line)
+        def match_target_seq_length(document, target_seq_length, filename,
+                                    line_idx, sentence_indices):
+            # match the target seq_length
+            for _ in range(10):
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: why is this 10?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 2]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
+            self.sentence_indices)
 
-            b_filename = a_filename
-            b_line = a_line
-            b_document = a_document
-        else:
-            # The rest of the time, B is a random sentence from the corpus
+        if self.sentence_indices[a_filename][a_line_idx] >= \
+                self.sentence_indices[a_filename][-1] or \
+                random.random() < 0.5:
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: can you please add this to the module's constructor parameters with 0.5 being the default?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 3]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
+
+            # About 50% of the time, B is a random sentence from the corpus
             label = 0
 
-            a_start_idx = random.randrange(len(a_document) - a_length)
+            for _ in range(10):
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: same here, why is it 10?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 4]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
+        def truncate_seq_pair(a_document, b_document, max_num_tokens):
+            # Truncates a pair of sequences to a maximum sequence length
+            while True:
+                total_length = len(a_document) + len(b_document)
+                if total_length <= max_num_tokens:
+                    break
+
+                trunc_document = a_document if len(
+                    a_document) > len(b_document) else b_document
+                assert len(trunc_document) >= 1
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: please don't use assert, raise an exception with a meaningful message instead

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 5]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
+
+                if random.random() < 0.5:
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: I don't understand what this part is doing, could you please elaborate. This 0.5 isn't related to the 0.5 above, right?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 6]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
         output_mask = np.array(output_mask, dtype=np.float32)
-        input_mask = np.ones(self.max_seq_length, dtype=np.float32)
+        input_mask = np.zeros(self.max_seq_length, dtype=np.float32)
+        input_mask[:len(input_ids)] = 1
 
         input_type_ids = np.zeros(self.max_seq_length, dtype=np.int)
-        input_type_ids[a_length + 2:seq_length + 3] = 1
+        input_type_ids[len(a_document) + 2:len(output_ids) + 1] = 1
+
+        while len(input_ids) < self.max_seq_length:
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: do we need to pad everything to the max_seq_length? If yes (I guess for perf reasons) should we also expose this a (default-on) parameter?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 7]
```diff
@@ -228,18 +286,46 @@ def mask_ids(self, ids):
...
-        for i in ids:
-            if random.random() < self.mask_probability:
-                output_mask.append(1)
+        for cand_index in cand_indexes:
+            if (random.random() < self.mask_probability) and \
+                  cand_index[0] != self.tokenizer.special_tokens["[CLS]"] and \
+                  cand_index[0] != self.tokenizer.special_tokens["[SEP]"]:
                 if random.random() < 0.8:
-                    masked_ids.append(self.tokenizer.special_tokens["[MASK]"])
+                    for cand_index_i in cand_index:
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: cand_index_i is unused

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 8]
```diff
@@ -228,18 +286,46 @@ def mask_ids(self, ids):
...
+                        output_mask.append(1)
+                        masked_ids.append(
+                            self.tokenizer.special_tokens["[MASK]"])
                 elif random.random() < 0.5:
-                    masked_ids.append(random.randrange(self.vocab_size))
+                    for cand_index_i in cand_index:
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: cand_index_i is unused

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 9]
```diff
@@ -228,18 +286,46 @@ def mask_ids(self, ids):
...
+                        output_mask.append(1)
+                        for _ in range(10):
```
Comment (okuchaiev) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: why these 10 attempts? can we do something like random_word = special_token and then sample until it isn't?

File: nemo/nemo/utils/exp_logging.py [hunk 10]
```diff
@@ -144,8 +144,9 @@ def get_tb_writer(self, tb_dir=None, exist_ok=True):
                 self.tb_writer = SummaryWriter(self.tb_dir)
             except ImportError:
                 self.tb_writer = None
-                self.logger.info('Not using TensorBoard.')
-                self.logger.info('Install tensorboardX to use TensorBoard')
+                # giving NoneType Error, so commenting out temporarily
```
Comment (okuchaiev) [inline nemo/nemo/utils/exp_logging.py]: please add check "if self.logger is not None" to guard it against NoneType error

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 11]
```diff
@@ -71,13 +72,10 @@ def find_newlines(contents):
...
 
-                        # Ensure the line has at least max_seq_length tokens
-                        if num_tokens >= max_seq_length:
-                            yield start - 1
-                            used_tokens += num_tokens
+                        yield new_start
 
-                        total_tokens += num_tokens
                         start = new_start + 1
+
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: if we no longer track total_tokens value, we should remove it entirely. but i think we should still track that number of statistics.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 12]
```diff
@@ -136,16 +135,14 @@ def __getitem__(self, idx, min_doc_length=16):
...
-        # in BERT paper
-        seq_length = self.max_seq_length - num_special_tokens
-
-        a_length = random.randrange(min_doc_length,
-                                    seq_length - min_doc_length + 1)
-        b_length = seq_length - a_length
+        max_num_tokens = self.max_seq_length - num_special_tokens
+        target_seq_length = max_num_tokens
+        if random.random() < self.short_seq_prob:
+            target_seq_length = random.randint(2, max_num_tokens)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: 1. 2 is really small. compared to 512, it'll increase the number of batches by 256. shouldn't we make the smallest value a bit bigger, like 128? we can also random from a normal distribution like N(256, 10) instead of a uniform.

2. can we still keep the comment TODO: make seq_length = 512 for the last 10% of epochs?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 13]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
 
                 # Read line, remove newline, and decode as UTF8
                 doc_text = f.readline()[:-1].decode("utf-8", errors="ignore")
-                document = [self.tokenizer.token_to_id("[CLS]")]
-                document.extend(self.tokenizer.text_to_ids(doc_text))
-                document.append(self.tokenizer.token_to_id("[SEP]"))
+                document = self.tokenizer.text_to_ids(doc_text)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: do we not insert special tokens anymore?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 14]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
+                    break
 
-        if random.random() < 0.5:
-            # 50% of the time, B is the sentence that follows A
-            label = 1
+            return document, line_idx
 
-            a_start_idx = random.randrange(len(a_document) - seq_length)
-            b_start_idx = a_start_idx + a_length
+        target_seq_length_a = int(round(target_seq_length * 0.6))
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: what's 0.6?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 13 above)
Comment (chiphuyen) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: ok i saw it below

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 15]
```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
...
 
-        return (np.array(input_ids),
-                input_type_ids,
-                input_mask,
-                np.array(output_ids),
-                output_mask,
-                label)
+        return np.array(input_ids), input_type_ids, np.array(
+            input_mask, dtype=np.float32), np.array(output_ids), np.array(
+            output_mask, dtype=np.float32), label
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: suggestion: use () to wrap around the return values so that each can be in its own line. just a style choice :)

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 1 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: just followed similarly to the original BERT implementation:
https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L279

10 is an arbitrary choice for now searching the data to match the target_seq_length.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 2 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: You mean, we want to have an input argument for random_next_proportion?

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 3 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: This is also following the original BERT implementation:
https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L275

I'll add a comment.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 6 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: I guess we do need to pad everything to the max_seq_length, because otherwise seq_length won't be the same for a batch - error gets raised.
This is especially so for longer seq_length like 512.

We won't process the padded ones because we know they are paddings.
Perf is only affected for this padding part (for loop).

Reference to the original BERT implementation:
https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L125

Once point though - 
In the original BERT implementation pre-processing is done before the pre-training, so the pre-training perf doesn't get affected by this.
We're doing this during the pre-training, so pre-training perf does get affected.
Not sure how they do in ROBERTa implementation - would be good to check and improve (along with the masking implementation to make sure same word never gets masked twice).

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 7 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: changed it to be 
`for _ in range(cand_index):`

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 8 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: changed it to be
`for _ in range(cand_index):`

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 9 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: again, imitated the original BERT implementation - it shouldn't go over 2 iterations most of the time, but 10 is there just to make sure. I'll put the comment in.
Certainly there may be more elegant and fast way to do this.
Using `while` loop like you said may be another way.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 12 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: `min_num=2` and `randint` was following the original BERT implementation:
https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L241

But that's a good idea to give users as argument to control, so we can experiment with.
I'll put it as comment.

Re. 3.: That's 2-phase pre-training for speed up and is controlled outside of this.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 15 above)
Comment (khcs) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: good idea :) I'll comment the action item as a TODO for now...

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 14 above)
Comment (chiphuyen) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: sounds good. thanks!

**Comment by okuchaiev** [comment]
could you please resolve the conflicts?
